### Preprocesado de datos

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("../data_sample/Datos.csv")
df

,provincia,zona,titulo,PrecioActual,PrecioAnterior,metros,habitaciones,ascensor,localizacion,planta,baños,tags,descripcion,Enlace
0,madrid,ciudad-lineal,"Piso en calle de San Marcelo, 22, Ventas, Madrid",355000,0,69,2.0,S,EXTERIOR,5ª,1,"VIVIENDA,LUMINOSO,VISTAS,REFORMADA,PORTERO,EXT...",Particular Vende vivienda totalmente reformada...,https://www.idealista.com/inmueble/106956987/
1,madrid,carabanchel,"Piso en calle Cabo Nicolás Mur, San Isidro, Ma...",149000,159000,91,3.0,N,EXTERIOR,1ª,0,"PISO,EXCLUSIVA,INMOBILIARIA,OPORTUNIDAD",Inmobiliarias Encuentro vende en exclusiva la ...,https://www.idealista.com/inmueble/106906044/
2,madrid,centro,"Piso en calle de Rodas, Lavapiés-Embajadores, ...",195000,0,36,1.0,S,NaN,2ª,0,"EXCLUSIVA,ESTUDIO",ESTUDIO EN PLENO CENTRO DE MADRIDSarago Servic...,https://www.idealista.com/inmueble/107306175/
3,madrid,usera,"Piso en calle de Ferroviarios, Almendrales, Ma...",195000,0,58,1.0,S,INTERIOR,BAJO,0,"VIVIENDA,HOGAR,FUNCIONAL","Esta acogedora vivienda, ubicada en una planta...",https://www.idealista.com/inmueble/106325171/
4,madrid,tetuan,"Dúplex en Bellas Vistas, Madrid",715000,750000,140,3.0,S,EXTERIOR,2ª,0,"TERRAZA,EXCLUSIVA,MODERNO","Maravilloso ATICO de reciente construcción, co...",https://www.idealista.com/inmueble/106627265/
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
11821,madrid,centro,"Piso en calle del Amparo, Lavapiés-Embajadores...",189000,0,36,2.0,N,INTERIOR,BAJO,0,"PISO,INMOBILIARIA,HOGAR,NUEVO",¡Descubre tu nuevo hogar en el corazón de Madr...,https://www.idealista.com/inmueble/107292488/
11822,madrid,centro,"Piso en calle Gran Vía, Chueca-Justicia, Madrid",2600000,0,245,2.0,S,EXTERIOR,2ª,0,"PISO,LUJO,EXCLUSIVO,EXTERIOR",Piso totalmente exterior ubicado entre el barr...,https://www.idealista.com/inmueble/103878128/
11823,madrid,tetuan,"Piso en calle del Capitán Blanco Argibay, Vald...",219000,225000,56,1.0,S,EXTERIOR,BAJO,0,"APARTAMENTO,METRO",¡IDEAL INVERSORES! Precioso apartamento en Val...,https://www.idealista.com/inmueble/107025446/
11824,madrid,carabanchel,"Piso en calle de Aceuchal, Vista Alegre, Madrid",165000,0,74,3.0,N,EXTERIOR,2ª,0,"OPORTUNIDAD,METRO","**IDEAL INVERSORES, NO DEJES PASAR ESTA OPORTU...",https://www.idealista.com/inmueble/106946149/


In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11826 entries, 0 to 11825
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   provincia       11826 non-null  object 
 1   zona            11826 non-null  object 
 2   titulo          11826 non-null  object 
 3   PrecioActual    11826 non-null  int64  
 4   PrecioAnterior  11826 non-null  int64  
 5   metros          11826 non-null  int64  
 6   habitaciones    11460 non-null  float64
 7   ascensor        11033 non-null  object 
 8   localizacion    10730 non-null  object 
 9   planta          10601 non-null  object 
 10  baños           11826 non-null  int64  
 11  tags            11664 non-null  object 
 12  descripcion     11761 non-null  object 
 13  Enlace          11826 non-null  object 
dtypes: float64(1), int64(4), object(9)
memory usage: 1.3+ MB


In [4]:
df.describe()

,PrecioActual,PrecioAnterior,metros,habitaciones,baños
count,1.182600e+04,1.182600e+04,11826.000000,11460.000000,11826.000000
mean,1.030501e+06,7.359704e+04,153.790039,2.847731,0.394047
std,1.237718e+06,3.639753e+05,766.217750,1.432402,0.882134
min,1.200000e+04,0.000000e+00,11.000000,1.000000,0.000000
25%,2.890000e+05,0.000000e+00,68.000000,2.000000,0.000000
50%,6.200000e+05,0.000000e+00,103.000000,3.000000,0.000000
75%,1.329000e+06,0.000000e+00,160.000000,3.000000,0.000000
max,2.300000e+07,8.450000e+06,75000.000000,20.000000,7.000000


In [5]:
df.isnull().sum().sort_values(ascending=False)

planta            1225
localizacion      1096
ascensor           793
habitaciones       366
tags               162
descripcion         65
metros               0
PrecioAnterior       0
PrecioActual         0
titulo               0
zona                 0
provincia            0
baños                0
Enlace               0
dtype: int64

### Primer análisis

1. COLUMNAS:
- provincia: Provincia del inmueble. Siempre es Madrid (es constante).
- zona: Categórica. Distrito al que pertenece el inmueble.
- titulo: Texto. Título del anuncio.
- PrecioActual: Numérica. Precio de venta actual. En euros. TARGET.
- PrecioAnterior: Numérica. Precio previo en euros.
- metros: Numérica. Superficie en metros cuadrados.
- habitaciones: Numérica. Número de habitaciones.
- ascensor: Categórica binaria. Indica si el inmueble tiene ascensor.
- localizacion: Categórica binaria. Indica la orientación de la vivienda. (exterior o interior)
- planta: Texto (TODO: analizar si puede ser ordinal). Planta del inmueble.
- baños: Numérica. Número de baños.
- tags: Texto en etiquetas separadas por comas. Etiquetas del anuncio.
- descripcion: Texto. Descripción del anuncio.
- Enlace: "Texto". URL del anuncio. Funciona como un ID.

2. LOS DATOS

- Dataset con 11.826 filas y 14 columnas
- Nulos en: planta, localizacion, ascensor, habitaciones, tags, descripcion.
- Los nulos de habitaciones, al ser NaN, pandas lo carga como un float en vez de un int.
- PrecioActual va desde 12.000 a 23.000.000 euros. Media muy desviada (1.030.501) respecto a la mediana (620.000). Asimetría fuerte.
- PrecioAnterior: Media de 73.597, pero los percentiles valen todos 0. Más del 75% de los datos son ceros.
- metros: Media de 154 respecto a la mediana de 103. Desviación estándar de 766. El máximo es 75.000, algo imposible para una vivienda. El mínimo son 11m2, pero ese es, lamentablemente, más razonable.
- habitaciones: de 1 a 20 y mediana 3. El máximo es "plausible".
- baños: mínimo 0, mediana 9 y percentil de 75 en 0 también. Una vivienda no puede tener 0 baños. Hay que investigar.  

Posiblemente los 0s en cualquier columna signifiquen "sin datos". De todas formas, hay que analizarlo y demostrarlo por si no es así.

Procedemos con el estudio de las categóricas. Hay 9 columnas de tipo object donde hay texto o categóricas.

In [6]:
# provincia es una columna constante. El único dato es Madrid. No nos servirá de nada
df.provincia.value_counts()

provincia
madrid    11826
Name: count, dtype: int64

In [7]:
# Podemos ver 21 zonas distintas. Casi todas las ofertas pertenecen al barrio de salamanca.
# El cual es conocido por ser bastante caro.
# La zona "centro" no parece ser un distrito como los demás.
df.zona.value_counts()

zona
barrio-de-salamanca    2018
centro                 1890
chamberi                796
chamartin               706
tetuan                  596
moncloa                 580
carabanchel             567
ciudad-lineal           537
puente-de-vallecas      526
retiro                  508
hortaleza               444
fuencarral              391
arganzuela              386
san-blas                335
latina                  331
villaverde              314
usera                   302
villa-de-vallecas       212
vicalvaro               193
moratalaz                97
barajas                  97
Name: count, dtype: int64

In [8]:
# Valores S y N
df.ascensor.value_counts()

ascensor
S    8419
N    2614
Name: count, dtype: int64

In [9]:
# Llama la atención los valores -1 y -2
df.planta.value_counts()

planta
1ª             2108
2ª             1779
3ª             1570
BAJO           1509
4ª             1266
5ª              816
6ª              574
7ª              289
ENTREPLANTA     163
SÓTANO          128
8ª              124
9ª              106
10ª              47
11ª              35
12ª              22
-1               19
13ª              16
14ª               8
15ª               7
17ª               6
16ª               2
-2                2
20ª               1
27ª               1
22ª               1
21ª               1
18ª               1
Name: count, dtype: int64

In [10]:
# 57 tags distintos

tags_extraidos = df["tags"].dropna().str.split(",").explode().str.strip() # El explode es para que en el value counts no me busque por listas.
tags_extraidos.value_counts()

tags
PISO                   6115
VIVIENDA               5723
EXTERIOR               2516
METRO                  2237
AMPLIO                 1995
TERRAZA                1922
REFORMADO              1880
OPORTUNIDAD            1832
EXCLUSIVA              1648
LUMINOSO               1436
HOGAR                  1316
ESPECTACULAR           1231
INMOBILIARIA           1205
ÁTICO                  1095
FINCA                  1094
LUJO                    986
VISTAS                  937
NUEVO                   930
EXCLUSIVO               874
REFORMADA               867
EQUIPADA                859
CASA                    851
PARQUE                  800
INTERIOR                792
GARAJE                  792
ESTRENAR                725
PISCINA                 721
FUNCIONAL               663
HALL                    659
MODERNO                 649
APARTAMENTO             638
SUITE                   615
PRESTIGIOSO             587
ELEGANTE                542
IMPRESIONANTE           541
ESTUDIO        

In [11]:
# Para extraer lo que hay en la columna de titulo hay que ver cómo se estructura y
# casualmente se estructura siempre igual.

# PRIMERO: Tipo inmueble.
# Son 9 tipos distintos
tipo_inmueble = df["titulo"].str.split(" en ").str[0].str.strip()
tipo_inmueble.value_counts()

titulo
Piso                           9551
Ático                           782
Dúplex                          435
Estudio                         339
Casa o chalet independiente     320
Chalet adosado                  213
Chalet pareado                  171
Chalet                           14
Casa rural                        1
Name: count, dtype: int64

In [12]:
# El barrio es un poquito más "tricky", pero sale bien
# Quitamos ", Madrid" del final
sin_madrid = df["titulo"].str.rsplit(",", n=1).str[0]

# El barrio es lo último
barrio = sin_madrid.str.rsplit(",", n=1).str[-1].str.strip()

# Si lleva "X en " (Piso en, Dúplex en... etc), nos quedamos con lo posterior al último " en "
barrio = barrio.str.rsplit(" en ", n=1).str[-1].str.strip()

barrio.value_counts()

titulo
Goya                    612
Recoletos               467
Malasaña-Universidad    451
Castellana              415
Lavapiés-Embajadores    366
                       ... 
El Pardo                  5
Valdecarros               5
Pavones                   5
Aeropuerto                4
Horcajo                   2
Name: count, Length: 139, dtype: int64

## PREPROCESADO

Primeras dos columnas "extra":  
Barrio y tipo de inmueble son las dos primeras columnas que irán al dataset.  
Se puede hacer ahora porque es una transformación fila a fila. No hay data leakage, el modelo no entrenaría con esto.

In [13]:
df["tipo_inmueble"] = df["titulo"].str.split(" en ").str[0].str.strip()

df["tipo_inmueble"].value_counts()

tipo_inmueble
Piso                           9551
Ático                           782
Dúplex                          435
Estudio                         339
Casa o chalet independiente     320
Chalet adosado                  213
Chalet pareado                  171
Chalet                           14
Casa rural                        1
Name: count, dtype: int64

In [14]:
# Mismo código que arriba, pero ahora lo metemos en la columna nueva

sin_madrid = df["titulo"].str.rsplit(",", n=1).str[0]
barrio = sin_madrid.str.rsplit(",", n=1).str[-1].str.strip()
df["barrio"] = barrio.str.rsplit(" en ", n=1).str[-1].str.strip()

df["barrio"].value_counts()

barrio
Goya                    612
Recoletos               467
Malasaña-Universidad    451
Castellana              415
Lavapiés-Embajadores    366
                       ... 
El Pardo                  5
Valdecarros               5
Pavones                   5
Aeropuerto                4
Horcajo                   2
Name: count, Length: 139, dtype: int64

#### COLUMNA BAÑOS

In [15]:
# baños tiene demasiadas columnas con 0
# lo que quiere decir que:
# Es un valor indicado a propósito
df["baños"].value_counts().sort_index()

baños
0    9311
1    1063
2     989
3     319
4      87
5      34
6      17
7       6
Name: count, dtype: int64

In [16]:
# Agrupamos y comparamos contra el precio. Las viviendas con supuestamente cero baños deberían ser baratísimas.
# Pero no es así, por lo que ese 0 no es que "tenga cero baños" sino que faltan datos.
# Hay que convertirlo a NaN sí o sí
df.groupby("baños")["PrecioActual"].median()

baños
0     598000.0
1     330000.0
2     725000.0
3    1335000.0
4    2450000.0
5    2100000.0
6    2500000.0
7    2375500.0
Name: PrecioActual, dtype: float64

## NOTA IMPORTANTE:
Aquí abajo dejo una celda con "df["sin_datos_baños"] = (df["baños"] == 0).astype(int)"  
No estoy seguro de si esto cuenta o no como un encoding. Es una operación fila a fila para una columna nueva, pero no tengo claro si estaría enseñando algo al modelo.  
Lo dejo comentado porque es importante tenerlo en cuenta.  


La otra, baños_limpio, no tiene ninguna imputación como tal. Cambiar un 0 por un NaN no calcula nada sobre las filas. El modelo no aprende ningún parámetro sobre el conjunto de datos.

In [17]:
# Columna para guardar la "ausencia de datos". Si era 0, entonces 
## df["sin_datos_baños"] = (df["baños"] == 0).astype(int)

# Copia de "baños" con el 0 convertido en NaN. No es una imputación. No entrena nada.
df["baños_limpio"] = df["baños"].replace(0, np.nan)

### OTRAS COLUMNAS CON CEROS

In [18]:
# Aunque lo habíamos visto arriba, un bucle sencillo para ver proporción de ceros

numericas = ["PrecioActual", "PrecioAnterior", "metros", "habitaciones", "baños"]

for col in numericas:
    n_ceros = (df[col] == 0).sum()
    porcentaje = round((df[col] == 0).mean() * 100, 1)
    print(f"{col}: {n_ceros} ceros ({porcentaje}%)")

PrecioActual: 0 ceros (0.0%)
PrecioAnterior: 10713 ceros (90.6%)
metros: 0 ceros (0.0%)
habitaciones: 0 ceros (0.0%)
baños: 9311 ceros (78.7%)


In [19]:
# La que queda es precioAnterior, que tiene una barbaridad de ceros.
# Lo más probable es que no tuvieran datos anteriores sobre el precio de esa vivienda.

# Primero vamos a ver las que sí nos dan datos comparándolas con el precio actual
con_valor = df[df["PrecioAnterior"] > 0]

bajadas = (con_valor["PrecioAnterior"] > con_valor["PrecioActual"]).sum()
subidas = (con_valor["PrecioAnterior"] < con_valor["PrecioActual"]).sum()
iguales = (con_valor["PrecioAnterior"] == con_valor["PrecioActual"]).sum()

print("Filas con PrecioAnterior:", len(con_valor))
print("Bajadas de precio (anterior > actual):", bajadas)
print("Subidas (anterior < actual):", subidas)
print("iguales:", iguales)

Filas con PrecioAnterior: 1113
Bajadas de precio (anterior > actual): 1113
Subidas (anterior < actual): 0
iguales: 0


### SOBRE ESTOS DATOS

Obviamente no son creíbles. El precio de la vivienda solo tiende a subir. Podemos sacar dos conclusiones respecto a esto: una mala y la otra peor.  

La mala: No hay datos fiables sobre el precio anterior de la vivienda y en este dataset se ha obtenido lo que se ha podido.  
La peor: Idealista instancia cada vivienda como un objeto nuevo cada vez que sube de precio. Sin embargo, no hace lo mismo con las bajadas de precio y por eso el 100% de los cambios de precio sobre LA MISMA VIVIENDA son de reducción. Son ofertas. Esto altera por completo las métricas y cualquier tipo de estudio. Así solo parece que las viviendas están estables y que, de vez en cuando, puedes encontrar ofertas.  

Sea como sea, lo que debemos de interpretar de esta columna entonces no es su literalidad, sino "si hay o no rebaja".

In [20]:
# Vamos a entender que esta columna en realidad nos indica si hay o no rebaja.
df["flag_rebaja"] = (df["PrecioAnterior"] > 0).astype(int)

In [21]:
# Y vamos a sacar otra columna para el porcentaje de rebaja. Igual es redundante, pero no está mal tenerla tampoco. Es una magnitud más que nos puede servir.
# Y además es un dato que se saca fila a fila. No hay imputación.
# TODO: Revisar si hay imputación por haber multiplicado por 100 y redondeado a 2.
diferencia = df["PrecioAnterior"] - df["PrecioActual"]

df["pct_rebaja"] = (np.where(df["PrecioAnterior"] > 0, diferencia / df["PrecioAnterior"], 0) * 100).round(2)

In [22]:
df[["flag_rebaja", "pct_rebaja"]]

,flag_rebaja,pct_rebaja
0,0,0.00
1,1,6.29
2,0,0.00
3,0,0.00
4,1,4.67
...,...,...
11821,0,0.00
11822,0,0.00
11823,1,2.67
11824,0,0.00


### COLUMNAS ASCENSOR Y LOCALIZACIÓN

In [23]:
# Tiene 793 nulos.
# Para ver si son "nulos de verdad" lo mejor es cruzarlos contra el tipo de inmueble.
# Hipótesis: igual son Casas, no pisos, donde no puedes poner que "no tiene ascensor" porque obviamente no es que no tenga, sino que no puede existir

df.ascensor.value_counts(dropna=False)

ascensor
S      8419
N      2614
NaN     793
Name: count, dtype: int64

In [24]:
# Porcentaje de ascensor nulo dentro de cada tipo de inmueble
# Tengo que crear una columna porque no encuentro otra manera de analizar esto
# Después hago un drop de esta columna y listo.
# ESTO ES SOLO PARA EL ANÁLISIS, PARA CONFIRMAR LA HIPÓTESIS O DESCARTARLA
df["ascensor_es_nulo"] = df["ascensor"].isna()
porcentaje_nulos = (df.groupby("tipo_inmueble")["ascensor_es_nulo"].mean() * 100).round(2)
porcentaje_nulos

tipo_inmueble
Casa o chalet independiente    100.00
Casa rural                     100.00
Chalet                         100.00
Chalet adosado                 100.00
Chalet pareado                 100.00
Dúplex                           0.00
Estudio                          5.60
Piso                             0.58
Ático                            0.00
Name: ascensor_es_nulo, dtype: float64

#### HIPÓTESIS CONFIRMADA:
Todas las casas o chalets tienen estos nulos.  
Algunos estudios (el 5.6%) tienen estos nulos.  
Muy pocos pisos (0.58%) tienen estos nulos.  

Lo suyo sería crear la columna de ascensor_limpio y estos nulos llamarlos NO_APLICA cuando son chalets o casas. Con los pisos se quedarían los nulos y hay dos posibilidades:  
Droppearlos si no son muchos (aunque un 5% es bastante)  
Meter un cuarto clasificador llamado DESCONOCIDO.

In [25]:
df.tipo_inmueble.value_counts()

tipo_inmueble
Piso                           9551
Ático                           782
Dúplex                          435
Estudio                         339
Casa o chalet independiente     320
Chalet adosado                  213
Chalet pareado                  171
Chalet                           14
Casa rural                        1
Name: count, dtype: int64

In [26]:
chalet_casa = [ "Casa o chalet independiente", "Chalet adosado", "Chalet pareado", "Chalet", "Casa rural"]

es_chalet_casa = df["tipo_inmueble"].isin(chalet_casa)
es_chalet_casa

0        False
1        False
2        False
3        False
4        False
         ...  
11821    False
11822    False
11823    False
11824    False
11825    False
Name: tipo_inmueble, Length: 11826, dtype: bool

In [27]:
# Sencillo: Creo la columna, compruebo si es True y además tiene el nulo en la columna, entonces "NO_APLICA"
# Después compruebo Si es false Y ADEMÁS tiene el nulo en la columna. Entonces "DESCONOCIDO"
df["ascensor_limpio"] = df["ascensor"]
df.loc[es_chalet_casa & df["ascensor"].isna(), "ascensor_limpio"] = "NO_APLICA"
df.loc[~es_chalet_casa & df["ascensor"].isna(), "ascensor_limpio"] = "DESCONOCIDO"  # Recordatorio de que en python/pandas la negación es con la virgulilla, no con exclamación
df.ascensor_limpio.value_counts(dropna=False)

ascensor_limpio
S              8419
N              2614
NO_APLICA       719
DESCONOCIDO      74
Name: count, dtype: int64

In [28]:
df.drop(columns="ascensor_es_nulo", inplace=True)

In [29]:
df

,provincia,zona,titulo,PrecioActual,PrecioAnterior,metros,habitaciones,ascensor,localizacion,planta,baños,tags,descripcion,Enlace,tipo_inmueble,barrio,baños_limpio,flag_rebaja,pct_rebaja,ascensor_limpio
0,madrid,ciudad-lineal,"Piso en calle de San Marcelo, 22, Ventas, Madrid",355000,0,69,2.0,S,EXTERIOR,5ª,1,"VIVIENDA,LUMINOSO,VISTAS,REFORMADA,PORTERO,EXT...",Particular Vende vivienda totalmente reformada...,https://www.idealista.com/inmueble/106956987/,Piso,Ventas,1.0,0,0.00,S
1,madrid,carabanchel,"Piso en calle Cabo Nicolás Mur, San Isidro, Ma...",149000,159000,91,3.0,N,EXTERIOR,1ª,0,"PISO,EXCLUSIVA,INMOBILIARIA,OPORTUNIDAD",Inmobiliarias Encuentro vende en exclusiva la ...,https://www.idealista.com/inmueble/106906044/,Piso,San Isidro,NaN,1,6.29,N
2,madrid,centro,"Piso en calle de Rodas, Lavapiés-Embajadores, ...",195000,0,36,1.0,S,NaN,2ª,0,"EXCLUSIVA,ESTUDIO",ESTUDIO EN PLENO CENTRO DE MADRIDSarago Servic...,https://www.idealista.com/inmueble/107306175/,Piso,Lavapiés-Embajadores,NaN,0,0.00,S
3,madrid,usera,"Piso en calle de Ferroviarios, Almendrales, Ma...",195000,0,58,1.0,S,INTERIOR,BAJO,0,"VIVIENDA,HOGAR,FUNCIONAL","Esta acogedora vivienda, ubicada en una planta...",https://www.idealista.com/inmueble/106325171/,Piso,Almendrales,NaN,0,0.00,S
4,madrid,tetuan,"Dúplex en Bellas Vistas, Madrid",715000,750000,140,3.0,S,EXTERIOR,2ª,0,"TERRAZA,EXCLUSIVA,MODERNO","Maravilloso ATICO de reciente construcción, co...",https://www.idealista.com/inmueble/106627265/,Dúplex,Bellas Vistas,NaN,1,4.67,S
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
11821,madrid,centro,"Piso en calle del Amparo, Lavapiés-Embajadores...",189000,0,36,2.0,N,INTERIOR,BAJO,0,"PISO,INMOBILIARIA,HOGAR,NUEVO",¡Descubre tu nuevo hogar en el corazón de Madr...,https://www.idealista.com/inmueble/107292488/,Piso,Lavapiés-Embajadores,NaN,0,0.00,N
11822,madrid,centro,"Piso en calle Gran Vía, Chueca-Justicia, Madrid",2600000,0,245,2.0,S,EXTERIOR,2ª,0,"PISO,LUJO,EXCLUSIVO,EXTERIOR",Piso totalmente exterior ubicado entre el barr...,https://www.idealista.com/inmueble/103878128/,Piso,Chueca-Justicia,NaN,0,0.00,S
11823,madrid,tetuan,"Piso en calle del Capitán Blanco Argibay, Vald...",219000,225000,56,1.0,S,EXTERIOR,BAJO,0,"APARTAMENTO,METRO",¡IDEAL INVERSORES! Precioso apartamento en Val...,https://www.idealista.com/inmueble/107025446/,Piso,Valdeacederas,NaN,1,2.67,S
11824,madrid,carabanchel,"Piso en calle de Aceuchal, Vista Alegre, Madrid",165000,0,74,3.0,N,EXTERIOR,2ª,0,"OPORTUNIDAD,METRO","**IDEAL INVERSORES, NO DEJES PASAR ESTA OPORTU...",https://www.idealista.com/inmueble/106946149/,Piso,Vista Alegre,NaN,0,0.00,N


### COLUMNA LOCALIZACIÓN

In [30]:
# Partimos de lo mismo que hicimos con baños
# La localización (exterior o interior) no aplica a las casas o chalets.
# Aplica muy poco a duplex, estudios, pisos y en muy menor medida a los áticos

df["localizacion_es_nulo"] = df["localizacion"].isna()
porcentaje_nulos_loc = (df.groupby("tipo_inmueble")["localizacion_es_nulo"].mean() * 100).round(2)
porcentaje_nulos_loc

tipo_inmueble
Casa o chalet independiente    100.00
Casa rural                     100.00
Chalet                         100.00
Chalet adosado                 100.00
Chalet pareado                 100.00
Dúplex                           0.46
Estudio                          7.08
Piso                             3.66
Ático                            0.13
Name: localizacion_es_nulo, dtype: float64

In [31]:
# Mismo criterio que en baños

df["localizacion_limpio"] = df["localizacion"]
df.loc[es_chalet_casa & df["localizacion"].isna(), "localizacion_limpio"] = "NO_APLICA"
df.loc[~es_chalet_casa & df["localizacion"].isna(), "localizacion_limpio"] = "DESCONOCIDO"

df["localizacion_limpio"].value_counts(dropna=False)

localizacion_limpio
EXTERIOR       9087
INTERIOR       1643
NO_APLICA       719
DESCONOCIDO     377
Name: count, dtype: int64

In [32]:
# La columna que no me sirve
df.drop(columns="localizacion_es_nulo", inplace=True)

### IMPORTANTE:
Para saber si droppear las filas de "DESCONOCIDO" en la tabla de localizacion_limpio, hace falta un EDA.  
No es mi competencia, por lo que dejo ahí los datos para que se analicen más adelante.

### COLUMNA PLANTA

In [33]:
# Esta columna tiene 1225 nulos
# Se comprueban igual que los anteriores, por si acaso están relacionados con las viviendas "individuales"
df.planta.value_counts(dropna=False)

planta
1ª             2108
2ª             1779
3ª             1570
BAJO           1509
4ª             1266
NaN            1225
5ª              816
6ª              574
7ª              289
ENTREPLANTA     163
SÓTANO          128
8ª              124
9ª              106
10ª              47
11ª              35
12ª              22
-1               19
13ª              16
14ª               8
15ª               7
17ª               6
16ª               2
-2                2
20ª               1
27ª               1
22ª               1
21ª               1
18ª               1
Name: count, dtype: int64

In [34]:
# Se puede ver el mismo patrón que antes
# No es necesario explicar
df["planta_es_nulo"] = df["planta"].isna()
porcentaje_nulos_planta = (df.groupby("tipo_inmueble")["planta_es_nulo"].mean() * 100).round(2)
porcentaje_nulos_planta

tipo_inmueble
Casa o chalet independiente    100.00
Casa rural                     100.00
Chalet                         100.00
Chalet adosado                 100.00
Chalet pareado                 100.00
Dúplex                           1.84
Estudio                          2.65
Piso                             4.67
Ático                            5.50
Name: planta_es_nulo, dtype: float64

In [35]:
# Lo mismo que en las anteriores
df["planta_limpio"] = df["planta"]
df.loc[es_chalet_casa & df["planta"].isna(), "planta_limpio"] = "NO_APLICA"
df.loc[~es_chalet_casa & df["planta"].isna(), "planta_limpio"] = "DESCONOCIDO"

df["planta_limpio"].value_counts(dropna=False)

planta_limpio
1ª             2108
2ª             1779
3ª             1570
BAJO           1509
4ª             1266
5ª              816
NO_APLICA       719
6ª              574
DESCONOCIDO     506
7ª              289
ENTREPLANTA     163
SÓTANO          128
8ª              124
9ª              106
10ª              47
11ª              35
12ª              22
-1               19
13ª              16
14ª               8
15ª               7
17ª               6
16ª               2
-2                2
20ª               1
27ª               1
22ª               1
21ª               1
18ª               1
Name: count, dtype: int64

In [36]:
# Igual que antes, droppeo la que no me sirve, que era solo para hacer la comprobación
df.drop(columns="planta_es_nulo", inplace=True)

In [37]:
# Sótano a -1 por ser duplicada
df["planta_limpio"] = df["planta_limpio"].replace("SÓTANO", "-1")

df["planta_limpio"].value_counts(dropna=False)

planta_limpio
1ª             2108
2ª             1779
3ª             1570
BAJO           1509
4ª             1266
5ª              816
NO_APLICA       719
6ª              574
DESCONOCIDO     506
7ª              289
ENTREPLANTA     163
-1              147
8ª              124
9ª              106
10ª              47
11ª              35
12ª              22
13ª              16
14ª               8
15ª               7
17ª               6
16ª               2
-2                2
20ª               1
27ª               1
22ª               1
21ª               1
18ª               1
Name: count, dtype: int64

### COLUMNA HABITACIONES

In [38]:
# 366 nulos que tiene pinta de que pueden ir a algún tipo de inmueble, como todo lo anterior
df.habitaciones.value_counts(dropna=False)

habitaciones
3.0     3768
2.0     3383
1.0     1624
4.0     1490
5.0      690
NaN      366
6.0      291
7.0      102
8.0       48
9.0       26
10.0      15
11.0      10
15.0       3
16.0       3
13.0       3
12.0       2
20.0       1
17.0       1
Name: count, dtype: int64

In [39]:
# A este paso, debería haber credo una función
# Mismo procedimiento que las columnas anteriores
df["habitaciones_es_nulo"] = df["habitaciones"].isna()
nulos_habitaciones = df.groupby("tipo_inmueble")["habitaciones_es_nulo"].sum()
nulos_habitaciones.sort_values(ascending=False)

tipo_inmueble
Estudio                        339
Ático                           15
Dúplex                          10
Casa o chalet independiente      1
Chalet pareado                   1
Chalet adosado                   0
Chalet                           0
Casa rural                       0
Piso                             0
Name: habitaciones_es_nulo, dtype: int64

In [40]:
# CAMBIOS: Aquí no hay datos en los estudios (claro, no tienen habitación, son solo eso, una habitación)
# Por tanto, sí podemos indicar un NO_APLICA o incluso mejor, ponerles un 0 porque si no perderíamos que sea numérica
df["habitaciones_limpio"] = df["habitaciones"]

# Le ponemos un 0 solo a los que son Estudio. Técnicamente no es imputar un valor y además es operación fila a fila
# No hay riesgo de data leakage, ni encoding ni nada.
es_estudio = df["tipo_inmueble"] == "Estudio"
df.loc[es_estudio & df["habitaciones"].isna(), "habitaciones_limpio"] = 0

In [41]:
# Quedan estas con NaNs, PERO al ser una numérica, la podemos dejar así
# Ya en el modelo, en el Pipeline, imputaremos datos. De momento, debería quedar así
df.loc[df["habitaciones_limpio"].isna(), "tipo_inmueble"].value_counts()

tipo_inmueble
Ático                          15
Dúplex                         10
Casa o chalet independiente     1
Chalet pareado                  1
Name: count, dtype: int64

In [42]:
df.drop(columns="habitaciones_es_nulo", inplace=True)

### COLUMNA METROS

In [43]:
# Esta vivienda de 75k metros cuadrados hay que mirarla. Esa y cualquiera que supere cantidades absurdas
df.metros.describe()

count    11826.000000
mean       153.790039
std        766.217750
min         11.000000
25%         68.000000
50%        103.000000
75%        160.000000
max      75000.000000
Name: metros, dtype: float64

In [44]:
# Miramos las filas con más de 10000 m2 a ver si hay alguna
metros_superior = df[df["metros"] > 10000]
metros_superior[["titulo", "PrecioActual", "metros", "habitaciones", "tipo_inmueble", "Enlace"]]

,titulo,PrecioActual,metros,habitaciones,tipo_inmueble,Enlace
6979,"Piso en calle de Emilio Ferrari, s/n, Pueblo N...",192000,75000,3.0,Piso,https://www.idealista.com/inmueble/107196960/
7997,"Chalet adosado en Palomas, Madrid",1495000,33175,7.0,Chalet adosado,https://www.idealista.com/inmueble/106105382/


In [45]:
# ELIMINAR
# Se eliminan porque son superficies imposibles y desconocemos el dato real.
# Cualquier imputación sería artificial. Para "imputar" estos datos, lo más correcto sería hacer un modelo predictivo con target en metros
# Y averiguar cuántos m2 corresponden a estos datos.
print("Filas antes:", len(df)) # Esto era solo para comprobar que la línea de abajo funcionaba bien
df = df[df["metros"] < 10000].copy()
print("Filas después:", len(df))

Filas antes: 11826
Filas después: 11824


In [46]:
# Hecho esto, podemos ver otro tipo de incongruencias pero sobre los € por metro cuadrado
euros_m2 = df["PrecioActual"] / df["metros"]

# Buscamos algún precio que esté muy muy por debajo del mercado
# Si los precios de Madrid a día de hoy no bajan de 2000€ el m2, teniendo en cuenta que este dataset debe ser algo anterior (no indica fecha)
# Mirando "a ojo" deberíamos estimar que todo aquello que esté por debajo de 500 €/m2 tiene que estar mal
# Fuentes del precio: https://www.bankinter.com/blog/finanzas-personales/comprar-casa-barata-madrid
sus = df[euros_m2 < 500]
sus

,provincia,zona,titulo,PrecioActual,PrecioAnterior,metros,habitaciones,ascensor,localizacion,planta,...,Enlace,tipo_inmueble,barrio,baños_limpio,flag_rebaja,pct_rebaja,ascensor_limpio,localizacion_limpio,planta_limpio,habitaciones_limpio
2714,madrid,puente-de-vallecas,"Estudio en San Diego, Madrid",114500,0,811,NaN,NaN,NaN,BAJO,...,https://www.idealista.com/inmueble/107028561/,Estudio,San Diego,NaN,0,0.0,DESCONOCIDO,DESCONOCIDO,BAJO,0.0
3683,madrid,san-blas,"Piso en calle Marmolistas, Arcos, Madrid",12000,0,61,2.0,N,NaN,NaN,...,https://www.idealista.com/inmueble/107240309/,Piso,Arcos,NaN,0,0.0,N,DESCONOCIDO,DESCONOCIDO,2.0
6669,madrid,san-blas,"Piso en calle Marmolistas, Arcos, Madrid",12000,0,61,2.0,N,NaN,NaN,...,https://www.idealista.com/inmueble/106859553/,Piso,Arcos,NaN,0,0.0,N,DESCONOCIDO,DESCONOCIDO,2.0
7750,madrid,puente-de-vallecas,"Piso en Entrevías, Madrid",161000,0,479,3.0,N,NaN,4ª,...,https://www.idealista.com/inmueble/101241888/,Piso,Entrevías,1.0,0,0.0,N,DESCONOCIDO,4ª,3.0


In [47]:
# Llamativo que un estudio tenga 811 metros cuadrados
# Las otras dos son duplicadas (no había caído hasta ahora en que la columna del enlace me iba a impedir ver duplicadas)
# Además, son viviendas que se venden en una fracción de la propiedad. No ayudan nada al dataset. Hay que investigar estos nuevos casos también.
# Esto es una categoría nueva que debo añadir: PROINDIVISO

# Sea como sea, estas cuatro no pueden entrar.
df = df[euros_m2 >= 500].copy()

In [48]:
# Tras ver ese estudio de 811 metros cuadrados, toca ver los metros por tipo de inmueble a ver si hay valores disparatados
df.groupby("tipo_inmueble")["metros"].describe()

,count,mean,std,min,25%,50%,75%,max
tipo_inmueble,,,,,,,,
Casa o chalet independiente,320.0,644.378125,398.502295,45.0,350.0,587.5,811.50,3015.0
Casa rural,1.0,20.000000,NaN,20.0,20.0,20.0,20.00,20.0
Chalet,14.0,678.928571,488.664501,62.0,266.5,683.5,989.50,1674.0
Chalet adosado,212.0,278.731132,147.534584,62.0,159.0,279.0,372.50,710.0
Chalet pareado,171.0,359.315789,141.696173,42.0,263.0,360.0,451.50,766.0
Dúplex,435.0,148.827586,104.011166,25.0,77.0,120.0,195.00,605.0
Estudio,338.0,46.038462,30.238248,12.0,30.0,40.0,54.00,300.0
Piso,9547.0,122.730491,85.424411,12.0,68.0,99.0,145.00,1139.0
Ático,782.0,154.558824,103.825994,11.0,84.0,128.5,192.75,697.0


Hay TRES valores absurdos:  
- Estudio de 605 m2
- Piso de 1139 m2
- Ático de 697m2

In [49]:
# Vamos tipo a tipo
estudios_grandes = df[(df["tipo_inmueble"] == "Estudio") & (df["metros"] > 100)]
print("Estudios de mas de 70 m2:", len(estudios_grandes))
estudios_grandes[["titulo", "PrecioActual", "metros", "descripcion", "Enlace"]]

Estudios de mas de 70 m2: 12


,titulo,PrecioActual,metros,descripcion,Enlace
626,"Estudio en calle de la Arquitectura, 234, Acac...",999999,280,"LOFT SINGULAR DE DISEÑO, VIVIENDA Y ESTUDIO/OF...",https://www.idealista.com/inmueble/106382458/
752,"Estudio en calle del Canal del Bósforo, 60, Ca...",380000,160,Se venden 2 Lofts completamente reformados a e...,https://www.idealista.com/inmueble/106844068/
1632,"Estudio en Fuente del Berro, Madrid",395000,130,Fantástico Loft exterior en Manuel BecerraEnge...,https://www.idealista.com/inmueble/99070395/
3687,"Estudio en Cuzco-Castillejos, Madrid",650000,140,Ocasion fabuloso piso la castellana luminosisi...,https://www.idealista.com/inmueble/107212672/
4201,"Estudio en calle del Padre Rubio, 13, Ventilla...",685000,180,"abstenerse inmobiliarias, se vende o alquila l...",https://www.idealista.com/inmueble/105934732/
4479,"Estudio en calle de Manuel Pombo Angulo, 18, S...",705000,140,Venta directa de Propietaria. Abstenerse agenc...,https://www.idealista.com/inmueble/106243786/
6556,"Estudio en 12 de Octubre-Orcasur, Madrid",157000,113,"Ampli pis a Madrid, perfecte per a parelles. L...",https://www.idealista.com/inmueble/102290951/
7808,"Estudio en calle Sierra Molina, 24, San Diego,...",185000,125,NaN,https://www.idealista.com/inmueble/102301211/
8295,"Estudio en calle Francisco Salas, Bellas Vista...",165000,118,Gran oferta de Piso en MADRID. Situado en CALL...,https://www.idealista.com/inmueble/102184945/
9725,"Estudio en calle de Alonso Núñez, Berruguete, ...",1200000,168,"Por favor contactar por escrito, no llamar.Pon...",https://www.idealista.com/inmueble/107115683/


#### HALLAZGO
Algunos (bastantes) estudios son LOFTS. No es que los metros cuadrados estén mal, sino que, como son espacios diáfanos, Idealista los cuenta como estudio.  
Una barbaridad, pero hay que indicarle al modelo que son algo distinto.

In [50]:
# La manera más sencilla es hacer una flag que indique si ese estudio es un loft
descripcion_lower = df["descripcion"].fillna("").str.lower()
df["flag_loft"] = descripcion_lower.str.contains("loft").astype(int)

df[df["flag_loft"] == 1]["tipo_inmueble"].value_counts()

tipo_inmueble
Piso       82
Dúplex     34
Estudio    27
Ático       4
Name: count, dtype: int64

In [51]:
# Llama la atención que haya tantos pisos. Lo normal es que, igual que en estudio, la gente habrá marcado "piso" para indicar su loft porque
# técnicamente son eso, pisos. Igual que un loft de dos plantas puede ser un dúplex.
# Por si acaso, hay que mirarlo

pisos_loft = df[(df["tipo_inmueble"] == "Piso") & (df["flag_loft"] == 1)]
pisos_loft[["titulo", "zona", "PrecioActual", "metros", "descripcion"]].head(20)

,titulo,zona,PrecioActual,metros,descripcion
143,"Piso en Niño Jesús, Madrid",retiro,1050000,199,Loft único y lujoso en Niño JesúsAuténtico lof...
157,"Piso en Sanchinarro, Madrid",hortaleza,249900,67,¡Descubre tu nuevo hogar en este encantador lo...
287,"Piso en calle de Fernández de la Hoz, 28, Alma...",chamberi,599900,80,NO COBRAMOS COMISIÓN AGENCIA DIRECTAMENTE PROP...
313,"Piso en calle de las Virtudes, Almagro, Madrid",chamberi,210000,38,Fantástico LOFT de 38m² completamente reformad...
715,"Piso en calle de Tomás López, Goya, Madrid",barrio-de-salamanca,525000,44,"Loft en venta en Calle Tomas López, Goya, Sala..."
824,"Piso en calle de Moratalla, Pinar del Rey, Madrid",hortaleza,229000,120,Pack de 2 Lofts Independientes – Oportunidad d...
944,"Piso en avenida de la Virgen del Carmen, Pinar...",hortaleza,229000,105,¡Descubre tu nuevo hogar o inversión perfecta!...
1016,"Piso en calle de la Princesa, Argüelles, Madrid",moncloa,497000,70,"MARCELI VENDE: Junto a Plaza de España, en Pla..."
1097,"Piso en Adelfas, Madrid",retiro,290000,42,Loft exterior en calle del Doctor EsquerdoEnge...
1381,"Piso en calle de Eloy Gonzalo, Trafalgar, Madrid",chamberi,369000,66,"Piso interior, tipo Loft en planta sótano con ..."


### OTRAS FLAGS

In [52]:
# Lista de flags que veo necesarias tras todo el preprocesado y limpieza

# Nuda propiedad:
df["flag_nuda_propiedad"] = descripcion_lower.str.contains("nuda propiedad").astype(int)

# Proindiviso:
df["flag_proindiviso"] = descripcion_lower.str.contains("proindiviso").astype(int)

# Subasta:
df["flag_subasta"] = descripcion_lower.str.contains("subasta").astype(int)

# Okupación
df["flag_okupada"] = descripcion_lower.str.contains("okupa|ocupad").astype(int)

# Alquilada:
df["flag_alquilada"] = descripcion_lower.str.contains("alquilad").astype(int)

flags_especiales = ["flag_nuda_propiedad", "flag_proindiviso", "flag_subasta", "flag_okupada", "flag_alquilada"]
df[flags_especiales].sum()

flag_nuda_propiedad     80
flag_proindiviso         5
flag_subasta            26
flag_okupada           271
flag_alquilada         305
dtype: int64

In [53]:
# Tengo dudas con la alquilada y okupada porque podría ser "libre de okupas" o "anteriormente okupada"
# Hay que mirar en la descripción
# EDIT: Parece que está todo bien.
okupa = df[df["flag_okupada"] == 1]
okupa[["titulo", "PrecioActual", "descripcion"]]

,titulo,PrecioActual,descripcion
36,"Piso en calle Azagra, 14, Pinar del Rey, Madrid",136000,NO TE COBRAMOS COMISION. Oportunidad sólo para...
91,"Piso en calle de Juan de Vera, 4, Delicias, Ma...",256400,******Oportunidad sólo para inversores. El act...
141,"Piso en calle de Belzunegui, Puerta Bonita, Ma...",119000,PISO OCUPADO POR PERSONA SIN JUSTO TÍTULO (Ant...
149,"Piso en calle del Bario, 18, Butarque, Madrid",221200,NO TE COBRAMOS COMISION. Oportunidad sólo para...
153,"Dúplex en Pau de Carabanchel, Madrid",195000,***INMUEBLE OCUPADO POR TERCERO SIN TÍTULO***O...
...,...,...,...
11529,"Piso en calle Peña de Oroel, 4, Casco Históric...",98200,"*** OKUPADA, SE VENDE LA PROPIEDAD., NO L APOS..."
11636,"Piso en calle Emilio Raboso, San Diego, Madrid",52400,Oportunidad sólo para inversores. El activo se...
11701,"Piso en calle Tomelloso, Almendrales, Madrid",141100,Oportunidad sólo para inversores. El activo se...
11808,"Piso en calle de Gregorio Ortiz, 13, Villaverd...",70900,******Oportunidad sólo para inversores. El act...


### IMPORTANTE
Algunas descripciones salen con "vivienda ocupada" con C y no con K. Es por eso que arriba, al crear la variable, indiquén okupa|ocupad

### PREPROCESADO DUPLICADAS
Arriba, en una de las celdas, se puede ver que una casa estaba duplicada aunque no nos salían filas duplicadas inicialmente. Esto es porque existe un ID "oculto" que es la url.  
Si una misma casa se ha subido dos veces, la URL es distinta y por tanto no parece una fila duplicada. Hay que mirarlas y tratarlas.

In [54]:
# Si coinciden con estos, es que son duplicadas al 100%
clave = ["titulo", "PrecioActual", "metros", "habitaciones", "planta", "baños", "ascensor", "localizacion", "zona"]

todas_duplicadas = df.duplicated(subset=clave, keep=False)
sobrantes = df.duplicated(subset=clave, keep="first")
print("Filas duplicadas:", todas_duplicadas.sum())
print("Cantidad de filas que quedan fuera:", sobrantes.sum())

Filas duplicadas: 1086
Cantidad de filas que quedan fuera: 636


In [55]:
len(df)

11820

In [56]:
df["n_ausentes"] = df.isna().sum(axis=1)
df = df.sort_values("n_ausentes")
df = df.drop_duplicates(subset=clave, keep="first").copy()
df = df.drop(columns="n_ausentes")

len(df)

11184

## DF PROCESADO:

In [57]:
df

,provincia,zona,titulo,PrecioActual,PrecioAnterior,metros,habitaciones,ascensor,localizacion,planta,...,ascensor_limpio,localizacion_limpio,planta_limpio,habitaciones_limpio,flag_loft,flag_nuda_propiedad,flag_proindiviso,flag_subasta,flag_okupada,flag_alquilada
2232,madrid,moratalaz,"Piso en calle de la Hacienda de Pavones, Fonta...",250000,0,63,2.0,S,EXTERIOR,BAJO,...,S,EXTERIOR,BAJO,2.0,0,0,0,0,0,0
2228,madrid,chamartin,"Piso en Bernabéu-Hispanoamérica, Madrid",4800000,0,463,4.0,S,EXTERIOR,2ª,...,S,EXTERIOR,2ª,4.0,0,0,0,0,0,0
2254,madrid,villa-de-vallecas,"Piso en avenida del Ensanche de Vallecas, Ensa...",420000,0,100,3.0,S,EXTERIOR,1ª,...,S,EXTERIOR,1ª,3.0,0,0,0,0,0,0
2274,madrid,barrio-de-salamanca,"Piso en Goya, Madrid",1459000,0,97,2.0,S,EXTERIOR,5ª,...,S,EXTERIOR,5ª,2.0,0,0,0,0,0,0
2266,madrid,barrio-de-salamanca,"Piso en Goya, Madrid",690000,0,60,2.0,S,INTERIOR,4ª,...,S,INTERIOR,4ª,2.0,0,0,0,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
8434,madrid,fuencarral,"Piso en calle de Joaquín Lorenzo, Peñagrande, ...",810000,0,131,3.0,NaN,NaN,NaN,...,DESCONOCIDO,DESCONOCIDO,DESCONOCIDO,3.0,0,0,0,0,0,0
3623,madrid,chamartin,"Casa o chalet independiente en Nueva España, M...",3800000,0,391,NaN,NaN,NaN,NaN,...,NO_APLICA,NO_APLICA,NO_APLICA,NaN,0,0,0,0,0,0
2511,madrid,hortaleza,"Chalet adosado en calle Somontín, 51, Apóstol ...",390000,0,125,3.0,NaN,NaN,NaN,...,NO_APLICA,NO_APLICA,NO_APLICA,3.0,0,0,0,0,0,0
8113,madrid,hortaleza,"Chalet pareado en Valdebebas - Valdefuentes, M...",775000,0,127,3.0,NaN,NaN,NaN,...,NO_APLICA,NO_APLICA,NO_APLICA,3.0,0,0,0,0,0,0


### LOS NULOS:
Aparecen nulos en las columnas, pero hay que matizar:
1. Las columnas originales no las he tocado. Tienen sus nulos naturales por eso mismo. Para ver la columna tratada, casi todos se llaman "_limpio" por ejemplo: ascensor_limpio.
2. Baños_limpio y habitaciones_limpio son casos especiales. Son nulos INTENCIONALES porque los voy a imputar en el pipeline del modelo. Todo eso es post-split.

In [58]:
df.isnull().sum()

provincia                 0
zona                      0
titulo                    0
PrecioActual              0
PrecioAnterior            0
metros                    0
habitaciones            360
ascensor                751
localizacion           1031
planta                 1166
baños                     0
tags                    157
descripcion              62
Enlace                    0
tipo_inmueble             0
barrio                    0
baños_limpio           8773
flag_rebaja               0
pct_rebaja                0
ascensor_limpio           0
localizacion_limpio       0
planta_limpio             0
habitaciones_limpio      27
flag_loft                 0
flag_nuda_propiedad       0
flag_proindiviso          0
flag_subasta              0
flag_okupada              0
flag_alquilada            0
dtype: int64

## SPLIT

In [59]:
from sklearn.model_selection import train_test_split

X = df.drop(columns="PrecioActual")
y = df["PrecioActual"]

In [60]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=df["zona"])

In [61]:
X_train.shape

(8947, 28)

In [62]:
y_train.shape

(8947,)

In [63]:
# Se guardan los conjuntos completos
train = X_train.copy()
train["PrecioActual"] = y_train

test = X_test.copy()
test["PrecioActual"] = y_test

# Y se guarda todo por fin en data_sample
train.to_csv("../data_sample/train.csv", index=False)
test.to_csv("../data_sample/test.csv", index=False)